### PR TITLE
Better multiplayer state

### PR DESCRIPTION
**Pull request by polycone**
### New features
- Multiplayer state management
- System-wide events dispatching
- Testing capabilities for mod operations
- Handling of game mode selection (multiplayer host/client, single player)
- Server name support
- Loading overlay status text
- Player profile

### Changes
- Network client and player entity segregation
- Commands have execution context
- Dependency container resolves only registered instances
- Cursor position update changed to event-based (no stored state in player data)
- Execution context has base context now and it can be overridden in a stacked manner
- Multiplayer coordinator replaced with several purpose-scoped controllers
- Ambiguous "disabled" multiplayer mode is removed

### Fixes
- #193
- #225

### CODE REVIEW

File: src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        this.multiplayer = multiplayer;
+
+        server.ClientDisconnected += OnClientDisconnected;
+        events.Subscribe<ClientInitializationRequestEvent>(OnClientInitializationRequested);
+
+        client.StateChanged += OnClientStateChanged;
+        events.Subscribe<GameStartedEvent>(OnGameStarted);
+        events.Subscribe<GameQuitEvent>(OnGameQuit);
+        events.Subscribe<CurrentPlayerInitializedEvent>(OnCurrentPlayerInitialized);
+        events.Subscribe<WorldSyncRequestedEvent>(OnWorldSaveRequested);
```
Comment (zuev93) [inline src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs:51]: Feels like it shouldn't be a member of PlayersManagementController

File: src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        MultiplayerGame multiplayer
+    ) {
+        this.client = client;
+        this.profileProvider = profileProvider;
+        this.server = server;
+        this.worldManager = worldManager;
+        this.multiplayer = multiplayer;
+
+        server.ClientDisconnected += OnClientDisconnected;
+        events.Subscribe<ClientInitializationRequestEvent>(OnClientInitializationRequested);
```
Comment (zuev93) [inline src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs:45]: Also feels slightly wrong about those 2 events being here.

And what is a purpose of identites map? Right now I don't see any read access to it (beside assert).

File: src/MultiplayerMod.Test/Multiplayer/LoadOverlayPatch.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using HarmonyLib;
+using JetBrains.Annotations;
+using MultiplayerMod.Multiplayer.UI;
+
+namespace MultiplayerMod.Test.Multiplayer;
```
[the remark this answers is not in the document]
Comment (polycone) [reply src/MultiplayerMod.Test/Multiplayer/LoadOverlayPatch.cs:5]: Not sure, it's a patch for these particular tests (what if we need some logic there, idk, will see later and move there if required).

File: src/MultiplayerMod.Test/Multiplayer/MultiplayerConnectionTests.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+using MultiplayerMod.Test.Environment.Unity;
+using NUnit.Framework;
+
+namespace MultiplayerMod.Test.Multiplayer;
+
+[TestFixture]
+public class MultiplayerConnectionTests {
+
+    [Test]
+    public void HostSelfConnection() {
```
[the remark this answers is not in the document]
Comment (polycone) [reply src/MultiplayerMod.Test/Multiplayer/MultiplayerConnectionTests.cs:22]: It's player connection tests as you can see further it asserts players state at various conditions. The test server and the client are just for pumping real commands, there are no server / client tests. 
Also I guess I know what you mean, and it bothers me as well, I think we need to design something better for mod <-> platform interaction, now the mod depends on details of a platform network implementation (connection state, for example).

File: src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs (hunk 1 above)
Comment (polycone) [reply src/MultiplayerMod/Multiplayer/CoreOperations/PlayersManagement/PlayersManagementController.cs:51]: Hm, maybe🤔

**Review by zuev93 — APPROVED** [review]

**Pull request merged**